Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include "ArgumentCoders.h"
+#include "Decoder.h"
+#include "Encoder.h"
+
+namespace TestWebKitAPI {
+
+struct EncodingCounter {
+    struct CounterValues {
+        unsigned encodingLValue { 0 };
+        unsigned encodingRValue { 0 };
+
+        void assertEqual(unsigned expectedLValueCount, unsigned expectedRValueCount)
+        {
+            ASSERT_EQ(expectedLValueCount, encodingLValue);
+            ASSERT_EQ(expectedRValueCount, encodingRValue);
+        }
+    };
+
+    EncodingCounter(CounterValues& counterValues)
+        : m_counterValues(counterValues)
+    { }
+
+    void encode(IPC::Encoder& encoder) const & {
+        encoder << uint32_t(0);
+        ++m_counterValues.encodingLValue;
+    }
+
+    void encode(IPC::Encoder& encoder) && {
+        encoder << uint32_t(0);
+        ++m_counterValues.encodingRValue;
+    }
+
+    static std::optional<EncodingCounter> decode(IPC::Decoder&) { return std::nullopt; }
+
+    CounterValues& m_counterValues;
+};
+
+TEST(ArgumentCoder, MoveSemantics)
```

Comment:
This could be
```
class ArgumentCoderTests : public testing::Test {
protected:
   void runTest(...) ...
}

TEST_F(ArgumentCoderTest, EncodeRValueOptional) {
 ....
}
...
```

This way if the test ever fails we know what fails immediately.
Otherwise it's a major debugging task that involves a backtrace to understand which assert fails.

(E.g. we write tests because we expect them to fail, and if a test can fail, then the failure should be understandable.)

Reply:
Moved the tests into a testing suite.